keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,829 @@
+
+All of the validation desired can be achieved by updating the CRDs to take advantage of the OpenApi v3 schema validation. In particular, the `oneOf` and `minLength` fields can be used.
+
+There is a desire for some fields to be immutable, which is not yet supported by CRDs. See the immutable fields [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190603-immutable-fields.md) for the latest updates. As of August 2020, the KEP is provisional and has no clear timeline for when immutable fields will come to CRDs.
+
+Note that we may want to consider other fields to be marked as immutable.
+
+#### VolumeSnapshot
+
+| Operation |            Field             |                                                                                                                                               Reason                                                                                                                                                | HTTP RCode |
+| :-------: | :--------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :--------: |
+|  CREATE   |         spec.Source          |                                                                                       Exactly one of PersistentVolumeClaimName (Dynamic) or VolumeSnapshotContentName (Pre-provisioned) should be specified.                                                                                        |    400     |
+|  UPDATE   |         spec.Source          |                             Immutable, no updates allowed. If the user has specified an incorrect source, they must delete and remake the snapshot. The webhook validation server will not be able to guarantee that only incorrect sources are allowed to be updated.                              |    400     |
+|  CREATE   | spec.VolumeSnapshotClassName | Must not be the empty string. Can be unset (to use the default snapshot class, if it is set. If the default snapshot class is not set or there is more than 1 default class, then the hook will allow the creation but the snapshot will fail.), or set to a non-empty string (the snapshot class). |    400     |
+|  UPDATE   | spec.VolumeSnapshotClassName |                                    Same restrictions as CREATE. We won’t restrict updating by making this field immutable (only applying the same restrictions as creation) but this field should only be changed by those who know exactly what they are doing.                                    |    400     |
```

Comment:
Can we add the reason for not restricting updates?

Reply:
@yuxiangqian would know what specific scenarios are valid for changing the class name.

Reply:
Can you add a TODO to resolve this by the time we implement the webhook? We should very clearly understand the use cases for supporting mutability of each field, and perhaps even document the use case in the CRD schema.

Reply:
one scenario was the following:
1. user creates a VS which points to a non-existing volume snapshot class (fat-fingered or whatever reason), in this case the volume snapshot controller will fail the create a snapshot.
2. user updates the volume snapshot class to point to a correct one.
IMO it sounds like a reasonable mutation scenario?
If it's a mutable field, then it makes sense to me not to list this entry here?

Reply:
What happens if a valid snapshot was created, but then the user changes the class to something different?

I guess there's a similar scenario where a user could delete and create a volumesnapshotclass with the same name, but completely different settings.

Reply:
Added a TODO issue https://github.com/kubernetes-csi/external-snapshotter/issues/362